File: tests/inputs/hello.jl
# Sample Julia file to exercise syntax highlighting, indents,
# and folds in vorto. Open with `vorto assets/samples/hello.jl`.

#=
Tiny program exercising typical Julia constructs:
modules, parametric structs, multiple dispatch, abstract types,
comprehensions, broadcasting, string interpolation, macros,
ternaries, ranges, and block comments like this one.
=#

module Greetings

export Person, greet, classify

const DEFAULT_GREETING = "Hello"

abstract type Animal end

struct Person{T<:Integer}
    name::String
    age::T
    tags::Vector{Symbol}
end

Person(name::String, age::Integer) = Person(name, age, Symbol[])

isadult(p::Person) = p.age >= 18

"Greet a person, optionally with a custom prefix."
function greet(p::Person, prefix::AbstractString = DEFAULT_GREETING)
    return "$prefix, $(p.name)!"
end

# Multiple dispatch: a different method for plain strings.
greet(name::AbstractString) = "$DEFAULT_GREETING, $name!"

function classify(n::Integer)
    if n < 0
        :negative
    elseif n == 0
        :zero
    else
        iseven(n) ? :positive_even : :positive_odd
    end
end

squares(range = 1:10) = filter(x -> x > 5, [x^2 for x in range])

function tag_lengths(words)
    return Dict(w => length(w) for w in words)
end

end # module Greetings

using .Greetings

# Script entry — runs with `julia hello.jl`.
alice = Person("Alice", 30, [:admin, :early_bird])
println(greet(alice))
println(greet(alice, "Hi"))
println(greet("Bob"))

for n in -2:5
    println("$n is $(classify(n))")
end

# Broadcasting and a comprehension.
doubled = (1:5) .^ 2
println("squares: ", squares())
println("doubled: ", doubled)
println("lengths: ", Greetings.tag_lengths(["alpha", "beta", "gamma"]))

File: tests/inputs/hello.scala
// Sample Scala file to exercise syntax highlighting, indents,
// and folds in vorto. Open with `vorto assets/samples/hello.scala`.

package demo

/** A tiny program exercising typical Scala constructs:
  *  case classes, enums, traits, pattern matching, given/using,
  *  for-comprehensions, and string interpolation.
  */

enum Classification:
  case Negative, Zero, PositiveEven, PositiveOdd

final case class Person(name: String, age: Int, tags: List[Symbol] = Nil):
  def isAdult: Boolean = age >= 18

trait Greeter[A]:
  def greet(subject: A): String

object Greetings:
  val DefaultGreeting = "Hello"

  given Greeter[Person] with
    def greet(p: Person): String = s"$DefaultGreeting, ${p.name}!"

  def classify(n: Int): Classification =
    n match
      case x if x < 0      => Classification.Negative
      case 0               => Classification.Zero
      case x if x % 2 == 0 => Classification.PositiveEven
      case _               => Classification.PositiveOdd

  def squares(xs: Seq[Int]): Seq[Int] =
    for
      x <- xs
      sq = x * x
      if sq > 4
    yield sq

@main def run(): Unit =
  val people = List(
    Person("Alice", 30, List(Symbol("admin"))),
    Person("Bob", 17),
  )
  val greeter = summon[Greeter[Person]]
  people.filter(_.isAdult).foreach(p => println(greeter.greet(p)))

  (-2 to 5).foreach(n => println(s"$n -> ${Greetings.classify(n)}"))
  println(s"squares: ${Greetings.squares(1 to 6)}")
